doc/designs/rbcd.md

```diff
@@ -0,0 +1,385 @@
+host` and `ipa service` commands.
+
+|Command                         |Description                                                 |
+|--------------------------------|------------------------------------------------------------|
+|service-add-delegation          |Add new resource delegation to a service                    |
+|service-allow-add-delegation    |Allow users, groups, hosts or host groups to handle a resource delegation of this service.|
+|service-disallow-add-delegation |Disallow users, groups, hosts or host groups to handle a resource delegation of this service.|
+|service-remove-delegation       |Remove resource delegation from a service|
+|host-add-delegation             |Add new resource delegation to a host|
+|host-allow-add-delegation       |Allow users, groups, hosts or host groups to handle a resource delegation of this host.|
+|host-disallow-add-delegation    |Disallow users, groups, hosts or host groups to handle a resource delegation of this host.|
+|host-remove-delegation          |Remove resource delegation from a host|
+
+### Implementation
+
+#### IPA API commands
+
+In both general constrained delegation and resource-based constrained
+delegation rules, targets, and delegation permission details are expressed with
+`memberPrincipal` attribute. Since IPA's standard `LDAPAddMember` and
+`LDAPRemoveMember` classes operate on DNs for the members, they need to be
+overridden to represent Kerberos principals. The principal might be not present
+in the IPA realm and thus cannot be represented as an LDAP object to which DN
+could be constructed.
+
+##### General constrained delegation
+
+In case of general constrained delegation a special handling is added to
+`LDAPAddMember` and `LDAPRemoveMember` classes to handle `memberPrincipal`.
+The add/remove methods assume, and require via asserts, that all members be a
+DN. `get_member_dns()` needs to ignore any `memberPrincipal` values and return
+only the DN-based values when adding targets to rules to let the standard
+mechanics of `LDAP*Member` do their work.
+
+In order to handle the `memberPrinicpal` values a `post_callback()` is required.
+This also means that there be at worst two writes per membership update.
+Given that this feature is not expected to be frequently used then speed and
+efficiency are not a factor.
+
+Similarly, we enforce that only a target is a member of a rule, and not
+another rule. That would be an undefined relationship. To do this each member
+needs to be retrieved and evaluated before adding as a member.
+
+A referential integrity rule is needed for ipaallowedtarget.
+
+A referential integrity rule is needed, but is not possible, for
+memberPrincipal. It is not possible because it is not a DN. This adds the
+potential to have dangling pointers.
+
+The ACL system also provides a means of limiting which user's a ticket may be
+obtained for using the `ipaAllowToImpersonate` attribute. This is not
+implemented.
```

Comment:
Is this a future feature or something that is possible but not necessary?

Reply:
We certainly might want to implement this as well but I don't see a realistic need for this right now. IPA delegation model is not really providing a clear separation and members of various permissions aren't considered totally separate admins. If we disallow use of members of `admins` group in RBCD, it might simply break the whole IPA setup.